src/java/org/apache/cassandra/db/Slice.java

```diff
@@ -200,7 +200,12 @@ public Slice forPaging(ClusteringComparator comparator, Clustering<?> lastReturn
             if (cmp < 0 || (inclusive && cmp == 0))
                 return this;
 
-            return new Slice(start, inclusive ? ClusteringBound.inclusiveEndOf(lastReturned) : ClusteringBound.exclusiveEndOf(lastReturned));
+            ClusteringBound<?> nextEnd = inclusive ? ClusteringBound.inclusiveEndOf(lastReturned) : ClusteringBound.exclusiveEndOf(lastReturned);
+
+            if (isEmpty(comparator, start, nextEnd))
```

Comment:
suggestion, not entirely sure it makes things clearer, but maybe bake this in the `cmp` check above - we can return null if `cmp == 0 && (!isInclusive || !start.isInclusive())` I think?